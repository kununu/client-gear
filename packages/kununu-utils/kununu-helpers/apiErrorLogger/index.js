import {logger} from '../../kununu-logger';
import isClientRender from '../isClientRender';

export default function (exception, application) {
  if (!isClientRender()) {
    const {status, url} = exception.response;

    logger.error({
      exception: {
        status,
        url,
      },
      message: 'API has reached an error',
      application,
    });
  }
}
