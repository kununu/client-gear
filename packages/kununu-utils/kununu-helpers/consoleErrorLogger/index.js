import {logger} from '../../kununu-logger';
import isClientRender from '../isClientRender';

export default function (error, label) {
  if (isClientRender()) {
    logger.log('error', {
      custom: true,
      error,
      label,
    });
  }
}
