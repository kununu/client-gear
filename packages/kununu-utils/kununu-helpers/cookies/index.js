import {serialize} from 'cookie';

const cookies = {
  get: name => document.cookie.split('; ').reduce((r, v) => {
    const parts = v.split('=');
    const cookie = parts[0] === name ? decodeURIComponent(parts[1]) : r;
    const shouldBeParsed = cookie[0] === '{' || cookie[0] === '[';
    return shouldBeParsed ? JSON.parse(cookie) : cookie;
  }, ''),
  set: (name, value, options) => {
    // allow you to work with cookies as objects.
    if (typeof value === 'object') {
      value = JSON.stringify(value);
    }

    const parsedOptions = typeof options === 'object' ?
      Object.keys(options).map(key => `${key}=${options[key]}`).join('; ') :
      null;

    document.cookie = serialize(name, value, parsedOptions);
  },
};

export default cookies;
