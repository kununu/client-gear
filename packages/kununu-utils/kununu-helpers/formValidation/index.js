export const EMPTY = 'EMPTY';
export const TOO_SHORT = 'TOO_SHORT';
export const NOT_VALID = 'NOT_VALID';

/**
 * Validates if a string is empty
 * @author Daniel Stein
 * @param {string} value [of input field]
 */
export function notEmpty (value) {
  if (!value) {
    return EMPTY;
  }

  return undefined;
}

/**
 * Validates if a value has a specified minimum length
 * @author Daniel Stein
 * @param {string} value [of input field]
 * @param {number} length
 */
function minLength (value, length) {
  if (value.trim().length < length) {
    return TOO_SHORT;
  }

  return undefined;
}

/**
 * Validates if a value has a minimum of 10 characters
 * @author Daniel Stein
 * @param {string} value [of input field]
 */
export const minLength10 = (value) => minLength(value, 10);

// Regex from https://stackoverflow.com/questions/46155/how-to-validate-email-address-in-javascript
const emailPattern = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
export function isEmail (value) {
  if (!emailPattern.test(value)) {
    return NOT_VALID;
  }

  return undefined;
}

export const validationTypes = {
  isEmpty: 'isEmpty',
  minLength: 'minLength',
  isEmail: 'isEmail',
};

/**
 * Checks a given value for each validator in a validations array
 *
 * @param {string} value
 * @param {Array} validations
 */
export function validateField (value, validations) {
  return validations.reduce((acc, validation) => {
    if (acc) return acc; // Always return error message

    switch (validation.type) {
      case validationTypes.isEmpty:
        if (notEmpty(value) !== undefined) return validation.message;
        break;
      case validationTypes.minLength:
        if (minLength(value, validation.minLength)) return validation.message;
        break;
      case validationTypes.isEmail:
        if (isEmail(value) !== undefined) return validation.message;
        break;
      default:
        break;
    }
    return false;
  }, false);
}
