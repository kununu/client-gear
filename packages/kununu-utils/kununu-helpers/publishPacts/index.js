const path = require('path');

const pact = require('@pact-foundation/pact-node');
const branchName = require('branch-name');
const Git = require('nodegit');

const packageJson = require(path.join(process.cwd(), 'package.json')); // eslint-disable-line import/no-dynamic-require

function isMaster (currentBranchName) {
  return currentBranchName === 'master';
}

function getReleaseCandidate (currentBranchName, hash) {
  return isMaster(currentBranchName) ? '' : `-rc.${hash.substr(0, 7)}`;
}

function getConsumerVersionByPackageJson (currentBranchName, hash) {
  return `${packageJson.version}${getReleaseCandidate(currentBranchName, hash)}`;
}

/**
 * Publishes pacts from a directory to a pact broker.
 * For convenience, this also finds out which git branch and
 * commit hash is currently needed for tagging and publishing
 *
 * @param {string} pactDir to pacts
 * @param {string} brokerUrl pacts broker url
 */
async function publishPacts (pactDir, brokerUrl) {
  // get current branch name
  const currentBranchName = await branchName.get();

  // get current hash of this git repo
  const repo = await Git.Repository.open(process.cwd());
  const commit = await repo.getHeadCommit();
  const hash = commit.sha();

  const consumerVersion = getConsumerVersionByPackageJson(currentBranchName, hash);

  // publish pacts
  const options = {
    pactFilesOrDirs: [pactDir],
    pactBroker: brokerUrl,
    consumerVersion,
    tags: [currentBranchName],
  };

  await pact.publishPacts(options);
}

module.exports = publishPacts;
