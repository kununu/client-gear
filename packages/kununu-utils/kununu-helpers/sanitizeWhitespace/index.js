/* eslint-disable import/prefer-default-export */
const sanitizeWhitespaceRegex = /\r\n|\r|\n| +/g;

export function sanitizeWhitespace (value) {
  return value.replace(sanitizeWhitespaceRegex, ' ');
}
