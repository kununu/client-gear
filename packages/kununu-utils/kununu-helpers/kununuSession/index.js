import {v4 as uuidv4} from 'uuid';

import cookies from '../cookies';
import isClientRender from '../isClientRender';

const KUNUNU_SESSION_ID_NAME = 'kununu_session_id';

const publishLoggedInEvent = (fetchApi) => {
  const params = {
    body: {
      type: 'login',
      source: 'logged_in_entrance',
    },
    credentials: 'same-origin',
    headers: {},
    method: 'POST',
  };

  fetchApi('/kunubi/event/logged-in-entrance', {}, params, 10000, false)();
};

const checkKununuSession = (config = {}) => {
  const {
    cookieName = KUNUNU_SESSION_ID_NAME,
    cookieValueGenerator = uuidv4,
    fetchApiFunc = () => {},
    shouldPublishLoggedInEvent = false,
  } = config;

  if (isClientRender()) {
    const sessionCookie = cookies.get(cookieName);
    const midnight = new Date();
    let expires = new Date();

    // sets the expiry date to midnight
    expires.setMinutes(expires.getMinutes() + 30);
    midnight.setHours(24, 0, 0, 0);

    if (expires > midnight) {
      expires = midnight;
    }

    const cookieProps = {
      expires,
      path: '/',
    };

    if (sessionCookie) {
      cookies.set(KUNUNU_SESSION_ID_NAME, sessionCookie, cookieProps);
    } else {
      cookies.set(KUNUNU_SESSION_ID_NAME, cookieValueGenerator(), cookieProps);
      if (shouldPublishLoggedInEvent && cookies.get('kununu_user_logged_info')) {
        publishLoggedInEvent(fetchApiFunc);
      }
    }
  }
};

export default checkKununuSession;
