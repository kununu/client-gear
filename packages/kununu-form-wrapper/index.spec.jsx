import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {shallow, mount} from 'enzyme'; // eslint-disable-line import/no-extraneous-dependencies
import {validationTypes} from '@kununu/kununu-utils/dist/kununu-helpers/formValidation';

import FormWrapper from './index';

describe('Returns current domain from a request object', () => {
  it('renders without crashing', () => {
    const MockComponent = () => (
      <input
        type="text"
        name="mock_input"
      />
    );

    const WrapperComponent = FormWrapper(MockComponent);

    const getInitialFieldsForUser = () => ({
      answer: {
        value: '',
        error: false,
        touched: false,
        validations: [
          {
            type: validationTypes.isEmpty,
            message: 'COMPANY_PROFILE_QA_ANSWER_AS_EMPLOYER',
          },
        ],
      },
    });

    const wrapper = shallow(<WrapperComponent getInitialFields={() => getInitialFieldsForUser()} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders async initial fields without crashing', () => {
    const MockComponent = () => (
      <input
        type="text"
        name="mock_input"
      />
    );

    const WrapperComponent = FormWrapper(MockComponent);

    let initialFields = () => ({});

    setTimeout(() => {
      initialFields = () => ({
        answer: {
          value: '',
          error: false,
          touched: false,
          validations: [
            {
              type: validationTypes.isEmpty,
              message: 'COMPANY_PROFILE_QA_ANSWER_AS_EMPLOYER',
            },
          ],
        },
      });
    }, 1000);

    const wrapper = shallow(<WrapperComponent getInitialFields={() => initialFields()} />);

    expect(wrapper).toMatchSnapshot();
  });
});

describe('possible interactions', () => {
  let WrapperComponent;
  let getInitialFieldsForUser;

  beforeEach(() => {
    localStorage.clear();

    /* eslint-disable react/require-default-props, react/destructuring-assignment */
    class MockComponent extends Component {
      static propTypes = {
        updateStateFromLocalStorage: PropTypes.func,
        updateLocalStorageFromState: PropTypes.func,
        handleUserInput: PropTypes.func,
        handleSubmit: PropTypes.func,
        touchForm: PropTypes.func,
        resetFormFields: PropTypes.func,
        testProp: PropTypes.string,
      };

      componentDidMount () {
        this.props.updateStateFromLocalStorage('some_key');
      }

      componentWillUpdate () {
        this.props.updateLocalStorageFromState('some_key');
      }

      onSubmit = () => {
        this.props.resetFormFields();
      }

      render () {
        return (
          <form onSubmit={e => this.props.handleSubmit(e, this.onSubmit)}>
            <input
              type="text"
              name={this.props.testProp}
              onChange={this.props.handleUserInput}
            />
            <button
              onClick={() => this.props.touchForm(true)}
              type="button"
            />
          </form>
        );
      }
    }
    /* eslint-enable react/require-default-props, react/destructuring-assignment */

    WrapperComponent = FormWrapper(MockComponent);

    getInitialFieldsForUser = () => ({
      answer: {
        value: '',
        error: false,
        touched: false,
        validations: [
          {
            type: validationTypes.isEmpty,
            message: 'value is way too short',
          },
        ],
      },
    });
  });


  it('validates input value when it confirms to validation condition', () => {
    const wrapper = shallow(<WrapperComponent getInitialFields={() => getInitialFieldsForUser()} />);
    const form = mount(wrapper.get(0));
    const input = form.find('input');

    input.simulate('change', {
      target: {
        value: 'dweldmd lwed wfwefewfwf',
        name: 'answer',
      },
    });

    expect(wrapper.state().fields.answer.error).toEqual(false);
  });

  it('throws correct error according to set validation conditions', () => {
    const wrapper = shallow(<WrapperComponent getInitialFields={() => getInitialFieldsForUser()} />);
    const form = mount(wrapper.get(0));
    const input = form.find('input');

    input.simulate('change', {
      target: {
        value: '',
        name: 'answer',
      },
    });

    expect(wrapper.state().fields.answer.error).toEqual('value is way too short');
  });

  it('will updates field values in localstorage on change', () => {
    const wrapper = shallow(<WrapperComponent getInitialFields={() => getInitialFieldsForUser()} />);
    const form = mount(wrapper.get(0));
    const input = form.find('input');
    const expectedInputValue = 'will be stored in localstorage';

    input.simulate('change', {
      target: {
        value: expectedInputValue,
        name: 'answer',
      },
    });

    // trigger componentWillUpdate
    form.setProps({testProp: 'some text here'});

    expect(localStorage.setItem).toHaveBeenLastCalledWith('some_key', JSON.stringify({fields: {answer: {value: expectedInputValue}}}));
  });

  it('will load field value from localstorage on mount', () => {
    const expectedInputValue = 'will be stored in localstorage';

    localStorage.setItem('some_key', JSON.stringify({fields: {answer: {value: expectedInputValue}}}));

    const wrapper = shallow(<WrapperComponent getInitialFields={() => getInitialFieldsForUser()} />);

    // still mount form, in order to trigger componentDidMount
    mount(wrapper.get(0));

    expect(wrapper.state().fields.answer.value).toEqual(expectedInputValue);
  });

  it('will touch fields when triggered', () => {
    const wrapper = shallow(<WrapperComponent getInitialFields={() => getInitialFieldsForUser()} />);
    const form = mount(wrapper.get(0));
    const input = form.find('input');
    const button = form.find('button');

    input.simulate('change', {
      target: {
        value: '123',
        name: 'answer',
      },
    });

    button.simulate('click');

    expect(wrapper.state().fields.answer.touched).toEqual(true);
  });

  it('will fire a callback defined in child component on submit and reset field values', () => {
    const wrapper = mount(<WrapperComponent getInitialFields={() => getInitialFieldsForUser()} />);
    const formElement = wrapper.find('form');
    const input = wrapper.find('input');

    input.simulate('change', {
      target: {
        value: 'some input text',
        name: 'answer',
      },
    });

    formElement.simulate('submit');

    expect(wrapper.state().fields.answer.value.length).toEqual(0);
  });

  it('will give you back props based on provided ID', () => {
    const wrapper = shallow(<WrapperComponent getInitialFields={() => getInitialFieldsForUser()} />);
    const form = mount(wrapper.get(0));
    const input = form.find('input');
    const expectedOutput = 'some input text';

    input.simulate('change', {
      target: {
        name: 'answer',
        value: expectedOutput,
      },
    });

    const propsFromWrapper = wrapper.props().getFormFieldProps('answer');

    expect(propsFromWrapper.id).toEqual('answer');
    expect(propsFromWrapper.input.value).toEqual('some input text');
  });

  it('will check if form is empty', () => {
    const wrapper = shallow(<WrapperComponent getInitialFields={() => getInitialFieldsForUser()} />);
    const form = mount(wrapper.get(0));
    const propsFromWrapperBefore = wrapper.props().formIsEmpty();

    expect(propsFromWrapperBefore).toEqual(true);

    const input = form.find('input');

    input.simulate('change', {
      target: {
        name: 'answer',
        value: 'Oh hi',
      },
    });

    const propsFromWrapperAfter = wrapper.props().formIsEmpty();

    expect(propsFromWrapperAfter).toEqual(false);
  });

  it('will handle different field value types', () => {
    const initialFields = () => ({
      question: {
        value: '',
        error: false,
        touched: false,
        validations: [],
      },
      answers: {
        value: [],
        error: false,
        touched: false,
        validations: [],
      },
    });

    const wrapper = mount(<WrapperComponent getInitialFields={() => initialFields()} />);
    const form = mount(wrapper.get(0));
    const input = wrapper.find('input');

    input.simulate('change', {
      target: {
        value: 'question1',
        name: 'question',
      },
    });

    expect(wrapper.state().fields.question.value).toEqual('question1');

    input.simulate('change', {
      target: {
        value: 'question2',
        name: 'question',
      },
    });

    expect(wrapper.state().fields.question.value).toEqual('question2');

    input.simulate('change', {
      target: {
        value: 'answer1',
        name: 'answers',
      },
    });

    expect(wrapper.state().fields.answers.value).toEqual(['answer1']);

    input.simulate('change', {
      target: {
        value: 'answer2',
        name: 'answers',
      },
    });

    expect(wrapper.state().fields.answers.value).toEqual(['answer1', 'answer2']);

    input.simulate('change', {
      target: {
        value: 'answer1',
        name: 'answers',
      },
    });

    expect(wrapper.state().fields.answers.value).toEqual(['answer2']);

    input.simulate('change', {
      target: {
        value: 'answer3',
        name: 'answers',
      },
    });

    form.simulate('submit');

    expect(wrapper.state().fields).toEqual({
      question: {
        value: 'question2',
        error: false,
        touched: false,
        validations: [],
      },
      answers: {
        value: ['answer2', 'answer3'],
        error: false,
        touched: false,
        validations: [],
      },
    });
  });

  it('will handle custom errors', () => {
    const wrapper = shallow(<WrapperComponent getInitialFields={() => getInitialFieldsForUser()} />);

    wrapper.props().handleCustomError('answer', 'custom error');
    expect(wrapper.state().fields.answer.error).toEqual('custom error');
  });
});
