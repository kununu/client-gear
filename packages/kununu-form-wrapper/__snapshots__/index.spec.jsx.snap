// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Returns current domain from a request object renders without crashing 1`] = `
<MockComponent
  fields={
    Object {
      "answer": Object {
        "error": false,
        "touched": false,
        "validations": Array [
          Object {
            "message": "COMPANY_PROFILE_QA_ANSWER_AS_EMPLOYER",
            "type": "isEmpty",
          },
        ],
        "value": "",
      },
    }
  }
  getFormFieldProps={[Function]}
  getInitialFields={[Function]}
  handleSubmit={[Function]}
  handleUserBlur={[Function]}
  handleUserInput={[Function]}
  resetFormFields={[Function]}
  touchForm={[Function]}
  updateLocalStorageFromState={[Function]}
  updateStateFromLocalStorage={[Function]}
/>
`;
